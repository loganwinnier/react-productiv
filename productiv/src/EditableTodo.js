import React, { useState } from "react";
import Todo from "./Todo";
import TodoForm from "./TodoForm";

/** Show editable todo item.
 *
 * Props
 * - todo
 * - update(): fn to call to update a todo
 * - remove(): fn to call to remove a todo
 *
 * EditableTodoList -> EditableTodo -> { Todo, TodoForm }
 */

function EditableTodo({ todo, update, remove }) {

  /** Toggle if this is being edited */
  function toggleEdit() {
    todo.isEditing = todo.isEditing ? false : true;
  }

  /** Call remove fn passed to this. */
  function handleDelete() {
    remove(todo.id);
  }

  /** Edit form saved; toggle isEditing and update in ancestor. */
  function handleSave(formData) {
    toggleEdit();
    update(formData);
  }

  const todoHtml = <div className="mb-3">
    <div className="float-end text-sm-end">
      <button
        className="EditableTodo-toggle btn-link btn btn-sm"
        onClick={toggleEdit}>
        Edit
      </button>
      <button
        className="EditableTodo-delBtn btn-link btn btn-sm text-danger"
        onClick={handleDelete}>
        Del
      </button>
    </div>
    <Todo todo={todo}/>
  </div>;

  const todoEditForm = <TodoForm initialFormData={todo} handleSave={handleSave}/>

  return (
    <div className="EditableTodo">
      {todo.isEditing ? todoEditForm : todoHtml}
    </div>
  );
}

export default EditableTodo;
