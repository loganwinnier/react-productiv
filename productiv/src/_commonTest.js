const TEST_TODO = {
  id: 1,
  title: "todo",
  description: "desc",
  priority: 1 }

const TEST_TODOS = [
  {
  id: 1,
  title: "todo-1",
  description: "desc-1",
  priority: 1
  },
  {
    id: 2,
    title: "todo-2",
    description: "desc-2",
    priority: 2
  },
  {
    id: 3,
    title: "todo-3",
    description: "desc-3",
    priority: 3
  }]

export { TEST_TODO, TEST_TODOS }
