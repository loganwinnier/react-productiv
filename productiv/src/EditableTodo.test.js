import React from "react";
import { fireEvent, render } from "@testing-library/react";
import EditableTodo from "./EditableTodo";
import { TEST_TODO } from "./_commonTest";

describe("EditableTodo component", function () {
    it("renders without crashing", function () {
        render(<EditableTodo todo={TEST_TODO} />);
    });

    it("contains title, description, priority if not editing", function () {
        const result = render(<EditableTodo todo={TEST_TODO} />);
        expect(result.queryByText(TEST_TODO.title)).toBeInTheDocument();
        expect(result.queryByText(TEST_TODO.description)).toBeInTheDocument();
        expect(result.queryByText(TEST_TODO.priority)).toBeInTheDocument();
    });

    it("contains editing form if editing", function () {
        const { queryByText, container } = render(<EditableTodo todo={TEST_TODO} />);
        expect(container.getElementsByClassName('NewTodoForm').length).toEqual(0);
        const button = queryByText('Edit');
        fireEvent.click(button);

        expect(container.getElementsByClassName('NewTodoForm').length).toEqual(1);
    });

    it("matches snapshot", function () {
        const { container } = render(<EditableTodo todo={TEST_TODO} />);
        expect(container).toMatchSnapshot();
    });
});