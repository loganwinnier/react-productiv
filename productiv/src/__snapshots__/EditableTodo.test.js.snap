// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableTodo component matches snapshot 1`] = `
<div>
  <div
    class="EditableTodo"
  >
    <div
      class="mb-3"
    >
      <div
        class="float-end text-sm-end"
      >
        <button
          class="EditableTodo-toggle btn-link btn btn-sm"
        >
          Edit
        </button>
        <button
          class="EditableTodo-delBtn btn-link btn btn-sm text-danger"
        >
          Del
        </button>
      </div>
      <div
        class="Todo"
      >
        <div>
          <b>
            todo
          </b>
           
          <small>
            1
          </small>
        </div>
        <div>
          <small>
            desc
          </small>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Todo component matches snapshot 1`] = `
<div>
  <div
    class="EditableTodo"
  >
    <div
      class="mb-3"
    >
      <div
        class="float-end text-sm-end"
      >
        <button
          class="EditableTodo-toggle btn-link btn btn-sm"
        >
          Edit
        </button>
        <button
          class="EditableTodo-delBtn btn-link btn btn-sm text-danger"
        >
          Del
        </button>
      </div>
      <div
        class="Todo"
      >
        <div>
          <b>
            todo
          </b>
           
          <small>
            1
          </small>
        </div>
        <div>
          <small>
            desc
          </small>
        </div>
      </div>
    </div>
  </div>
</div>
`;
